pkg/fileutils/filelock.go

```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright The Dragonfly Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package fileutils
+
+import (
+	"fmt"
+	"os"
+	"sync"
+	"syscall"
+
+	"github.com/pkg/errors"
+)
+
+// FileLock defines a file lock implemented by syscall.Flock
+type FileLock struct {
+	fileName string
+	fd       *os.File
+	mutex    sync.Mutex
+}
+
+// NewFileLock create a FileLock instance
+func NewFileLock(name string) *FileLock {
+	return &FileLock{
+		fileName: name,
+	}
+}
+
+// Lock locks file.
+// If the file is already locked, the calling goroutine blocks until the file is unlocked.
+// If lock has been invoked without unlock, lock again will return an error.
+func (l *FileLock) Lock() error {
+	var (
+		fd  *os.File
+		err error
+	)
+	if fd, err = os.Open(l.fileName); err != nil {
+		return err
+	}
+
+	l.mutex.Lock()
+	if l.fd != nil {
+		l.mutex.Unlock()
+		return fmt.Errorf("invoke Unlock() before lock %s again", l.fileName)
+	}
+	l.fd = fd
+	l.mutex.Unlock()
+
+	if err := syscall.Flock(int(l.fd.Fd()), syscall.LOCK_EX); err != nil {
```

Comment:
The `l.fd` may be `nil`:

```
1. Lock:                   | Unlock:                  
2.   l.mutex.Unlock()      |   ...
3.                         |   l.mutex.Lock()
4.                         |   l.fd = nil
5.   syscall.Flock(l.fd)   |   ...
6.                         |   ...
```

I think that you can just ignore the thread-safe feature of `filelock` because that the `StartPeerServerProcess` will not be invoked in multi goroutines.